sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2230,6 +2230,8 @@ class Analyzer(
           val result = resolved transformDown {
             case UnresolvedMapObjects(func, inputData, cls) if inputData.resolved =>
               inputData.dataType match {
+                case ArrayType(et, false) if cls.isEmpty &&
+                  CatalystTypeConverters.isPrimitive(et) => Cast(inputData, inputData.dataType)
```

Comment:
if the expected array element type is same as the actual element array type, and the cutomer collection class is empty, then we can say that this `MapObjects` is a no-op and we can remove it.

The information of expected array element type is actually hidden in the map function, so I think it's hard to do the check here. How about we add an optimizer rule that if `MapObjects.lambdaFunction` is `LambdaVariable`, we eliminate this `MapObjects`